File: src/julia/lorenz_attractor.jl
#=
Lorenz-Attractor
https://docs.juliaplots.org/latest/user_gallery/misc/gr_lorenz_attractor/#Lorenz-Attractor
=#

# define the Lorenz attractor
Base.@kwdef mutable struct Lorenz
    dt::Float64 = 0.02
    σ::Float64 = 10
    ρ::Float64 = 28
    β::Float64 = 8 / 3
    x::Float64 = 1
    y::Float64 = 1
    z::Float64 = 1
end

function step!(l::Lorenz)
    dx = l.σ * (l.y - l.x)
    dy = l.x * (l.ρ - l.z) - l.y
    dz = l.x * l.y - l.β * l.z
    l.x += l.dt * dx
    l.y += l.dt * dy
    l.z += l.dt * dz
    return nothing
end

function generate_points()::DataFrame
    attractor = Lorenz()
    df = DataFrame(:x => Float64[], :y => Float64[], :z => Float64[])
    for _ in 1:500
        step!(attractor)
        x = attractor.x
        y = attractor.y
        z = attractor.z
        push!(
            df,
            Dict(:x => x, :y => y, :z => z),
        )
    end
    return df
end

File: src/python/lorenz_attractor.py
"""
Lorenz-Attractor
https://docs.juliaplots.org/latest/user_gallery/misc/gr_lorenz_attractor/#Lorenz-Attractor
"""

from pydantic import BaseModel
import pandas as pd


class Lorenz(BaseModel):
    dt: float = 0.02
    sigma: float = 10
    rho: float = 28
    beta: float = 8 / 3
    x: float = 1
    y: float = 1
    z: float = 1

    def step(self):
        dx = self.σ * (self.y - self.x)
        dy = self.x * (self.ρ - self.z) - self.y
        dz = self.x * self.y - self.β * self.z
        self.x += self.dt * dx
        self.y += self.dt * dy
        self.z += self.dt * dz


def generate_points(attractor: Lorenz = Lorenz()):
    df = pd.DataFrame(columns=["x", "y", "z"])
    for _ in range(500):
        attractor.step()
        x = attractor.x
        y = attractor.y
        z = attractor.z
        df.append({"x": x, "y": y, "z": z})
    return df
